File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
from heapq import heappush, heappop
from collections import deque

def solution(book_time):
    answer = 0
    book_time.sort()
    newtime = []
    N = len(book_time)
    for i in range(N):
        tmp = []
        for j in range(2):
            time = 0
            h, m = book_time[i][j].split(":")
            time += int(h) * 60 + int(m)
            tmp.append(time)
        newtime.append(tmp)
    heap = []
        
    for i in range(N):
        answer = max(answer, len(heap))
        if not heap:
            heappush(heap, (newtime[i][1], newtime[i][0]))
        else:
            tmp = heappop(heap)
            if newtime[i][0] >= tmp[0]+10:
                heappush(heap, (newtime[i][1], newtime[i][0]))
            else:
                heappush(heap, tmp)
                heappush(heap, (newtime[i][1], newtime[i][0]))
    answer = max(answer, len(heap))
    return answer

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;
class Solution {
    public int solution(String[][] book_time) {
        int answer = 0;
        Arrays.sort(book_time, (o1, o2) -> o1[0].compareTo(o2[0]));
        int N = book_time.length;
        int[][] newTime = new int[N][2];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < 2; j++) {
                int time = 0;
                String[] tmp = book_time[i][j].split(":");
                time += Integer.parseInt(tmp[0]) * 60;
                time += Integer.parseInt(tmp[1]);
                newTime[i][j] = time;
            }
        }
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (int i = 0; i < N; i++) {
            if (!pq.isEmpty()) {
                int tmp = pq.poll();
                if (newTime[i][0] < tmp + 10) {
                    pq.offer(tmp);
                }
            }
            pq.offer(newTime[i][1]);
            answer = Math.max(answer, pq.size());
        }
        
        return answer;
    }
}